src/DataStructures/Tensor/EagerMath/Magnitude.hpp

```diff
@@ -65,7 +66,9 @@ void magnitude(const gsl::not_null<Scalar<DataType>*> magnitude,
                             index_list<change_index_up_lo<Index>,
                                        change_index_up_lo<Index>>>& metric) {
   dot_product(magnitude, vector, vector, metric);
-  get(*magnitude) = sqrt(get(*magnitude));
+  // Spacetime dot product can be negative, so take the absolute value to define
+  // the magnitude
+  get(*magnitude) = sqrt(abs(get(*magnitude)));
```

Comment:
just a quick comment here, haven't looked at anything else.

We do a lot of magnitude calculations, so I would prefer to have a separate function that does the "extra math". Taking an unnecessary absolute value at every element boundary is a lot of overhead for no reason.

You could just check if the `Index` is spatial or spacetime, then decide based on that :shrug: